fix (#18501): chansend sending lines to terminal buffer in reverse order in Windows

3N4N opened the pull request:
Fixes #18501 

Problem: `chansend` in Windows sends lines in reverse order.
Cause: The newline character used was \n instead of \r\n.
Solution: If in Windows, adjust the newline characters.

3N4N commented on src/nvim/channel.c:
```diff
@@ -517,6 +517,45 @@ uint64_t channel_from_stdio(bool rpc, CallbackReader on_output, const char **err
+  size_t c = 0;
+  char *p = *str;
+
+  while (*p++) {
```
Should I use `strcnt`?
https://github.com/neovim/neovim/blob/778541067a2f0ec3ef4bc1b07def646323e852a9/src/nvim/memory.c#L269-L286

3N4N commented on src/nvim/channel.c:
```diff
@@ -517,6 +517,45 @@ uint64_t channel_from_stdio(bool rpc, CallbackReader on_output, const char **err
+  size_t c = 0;
+  char *p = *str;
+
+  while (*p++) {
+    c += *p == '\n' ? 1 : 0;
+  }
```
Should I use `strcnt`?
https://github.com/neovim/neovim/blob/778541067a2f0ec3ef4bc1b07def646323e852a9/src/nvim/memory.c#L269-L286

3N4N commented on src/nvim/channel.c:
```diff
@@ -517,6 +517,45 @@ uint64_t channel_from_stdio(bool rpc, CallbackReader on_output, const char **err
+  xstrlcat(ret, "\r\n", len);
+  while ((tok = os_strtok(NULL, "\n", &con))) {
+    xstrlcat(ret, tok, len);
+    xstrlcat(ret, "\r\n", len);
```
(Leaving this for myself when I get home) If the last character in `*str` isn't `\n`, then this will add an extra two characters. The value of `c` may then be incorrect.

3N4N commented on test/functional/ui/output_spec.lua:
```diff
@@ -55,7 +55,7 @@ describe("shell command :!", function()
       pending('FIXME #10804')
     end
-    child_session.feed_data((":!%s REP 30001 foo\n"):format(testprg('shell-test')))
+    child_session.feed_data((":!%s REP 30001 foo\r"):format(testprg('shell-test')))
```
I ended up changing the test itself. I mean, `<CR>` means Carriage Return. So `\r` seems more appropriate, right? I donno more in-depth technicalities of these end-of-line stuff, tho.

justinmk replied:
yes that's fine.

justinmk commented on src/nvim/channel.c:
```diff
@@ -517,6 +517,46 @@ uint64_t channel_from_stdio(bool rpc, CallbackReader on_output, const char **err
   return channel->id;
 }
 
+size_t newline_adjust(char **str)
```
- given your work since this PR, do you see any way to simplify this function? e.g. we were able to avoid strktok in the other PRs.
- needs a brief docstring explaining what's going on here, with a simple concrete example like: before "...", after: "..."

justinmk commented on src/nvim/channel.c:
```diff
@@ -563,6 +603,10 @@ size_t channel_send(uint64_t id, char *data, size_t len, bool data_owned, const
   }
 
+  if (chan->term) {
+    len += newline_adjust(&data);
```
instead of doing this here, can we find where `data` is prepared and change its logic? similar to the simple change from https://github.com/neovim/neovim/pull/18723

3N4N replied:
> So we could send a flag to `save_tv_as_string` that changes its separator char from `\n` to `\r\n` ?

That's what I did at first (see updates 4 and 5 in this [comment][1]). But `save_tv_as_string` handles putting text in general, so converting `\n` to `\r\n` will convert them when putting text in non-terminal buffers, too, which is undesired. We _could_ send a flag to determine if it's for a term channel, but: 1) I thought changing parameters of an existing function was too much of a hassle for a newbie in the codebase, and 2) I couldn't figure out if I should use `id`[2] or something else to figure out `chan->term` [3]. It's kind of about code design. I didn't wanna make big changes by shifting code around. Need guidance.

> `nvim_chan_send`, which only accepts String. I don't think we should be modifying raw string input--it's up to the caller to decide what raw data they want to send, right?

I actually didn't even see this. If we handle `\n` -> `\r\n` in `save_tv_as_string`, then this would be automatically fixed.

---

Sorry about adding links weirdly. GH is processing my links, prepending other links and stuff. Donno what that's about. Thought it best to paste it as text.


[1]: https://github.com/neovim/neovim/issues/18501#issuecomment-1174106059
2: https://github.com/3N4N/neovim/blob/85754dcdd992/src/nvim/eval/funcs.c#L873
3: https://github.com/neovim/neovim/blob/85754dcdd992/src/nvim/channel.c#L586

3N4N replied:
@justinmk I pushed a second commit (we can squash it later; I thought we should keep the changes available until merged) following your suggestion: only converting LF to CRLF when joining list items.

3N4N commented on test/functional/ui/output_spec.lua:
```diff
@@ -55,7 +55,7 @@ describe("shell command :!", function()
       pending('FIXME #10804')
     end
-    child_session.feed_data((":!%s REP 30001 foo\n"):format(testprg('shell-test')))
+    child_session.feed_data((":!%s REP 30001 foo\r"):format(testprg('shell-test')))
```
@3N4N remember to revert this change.

justinmk commented:
> But save_tv_as_string handles putting text in general, so converting \n to \r\n will convert them when putting text in non-terminal buffers, too, which is undesired. 

Can we add a `crlf` paramter to `save_tv_as_string` which is set to true by `f_chansend`?

3N4N commented:
> Can we add a crlf paramter to save_tv_as_string which is set to true by f_chansend?

Did something similar.

```
char *save_tv_as_string(typval_T *tv, ptrdiff_t *const len, bool endnl, bool is_chanterm)
  FUNC_ATTR_MALLOC FUNC_ATTR_NONNULL_ALL
{
#ifdef UNIX
  bool crlf = false;
#else
  bool crlf = is_chanterm ? true: false;
#endif
```
But I can do crlf as parameter directly as well, if you want.

3N4N commented on src/nvim/eval.c:
```diff
@@ -6177,9 +6177,15 @@ bool read_blob(FILE *const fd, blob_T *const blob)
 /// @returns an allocated string if `tv` represents a VimL string, list, or
 ///          number; NULL otherwise.
-char *save_tv_as_string(typval_T *tv, ptrdiff_t *const len, bool endnl)
+char *save_tv_as_string(typval_T *tv, ptrdiff_t *const len, bool endnl, bool is_chanterm)
   FUNC_ATTR_MALLOC FUNC_ATTR_NONNULL_ALL
 {
+#ifdef UNIX
+  bool crlf = false;
+#else
+  bool crlf = is_chanterm ? true: false;
+#endif
```
> Can we add a crlf paramter to save_tv_as_string which is set to true by f_chansend?

Did something similar^. But I can do crlf as parameter directly as well, if you want.

justinmk replied:
This assumes that no other callers will call `save_tv_as_string`. Seems better to make it explicit?

3N4N replied:
> Seems better to make it explicit?

True. I committed and pushed.

Can't figure out the EOF crash, though. If you had any idea.

3N4N replied:
Oh wait. It passed those tests now! Huh.

justinmk commented on src/nvim/eval.c:
```diff
@@ -6234,20 +6235,23 @@ char *save_tv_as_string(typval_T *tv, ptrdiff_t *const len, bool endnl)
   }
 
-  char *ret = xmalloc((size_t)(*len) + endnl);
+  char *ret = xmalloc((size_t)(*len) + (size_t)endnl * (crlf ? 2 : 1));
```
Seems like a mistake that we were adding `endnl` (a boolean) here? Instead, it should be used as a ternary which evaluates to an int literal?

3N4N replied:
I thought it was intentional. False evaluating to 0, etc.

justinmk replied:
If we change `endnl` to a int, this line can remain as it was.

justinmk replied:
> False evaluating to 0, etc.

maybe, but technically there are some places where we cast int to `bool`. Doesn't matter in this case but it seems unnecessarily clever. In any case, it's moot because it seems like a good idea to change the `endnl` to `int` instead of adding a new `crlf` param.

3N4N replied:
I'm a little unclear on turning `endnl` into an int. Wrote my confusion in the comments below. Meanwhile, I committed your ternary op suggestion. You're right. At least before it made sense in that it shaved off two/three characters of code. In my case, it was just idiotic.

justinmk commented on src/nvim/eval.c:
```diff
@@ -6175,9 +6175,10 @@ bool read_blob(FILE *const fd, blob_T *const blob)
 /// @returns an allocated string if `tv` represents a VimL string, list, or
 ///          number; NULL otherwise.
-char *save_tv_as_string(typval_T *tv, ptrdiff_t *const len, bool endnl)
+char *save_tv_as_string(typval_T *tv, ptrdiff_t *const len, bool endnl, bool crlf)
```
could change the `endnl` param to a `TriState` or `int` where:

```
0 = none
1 = LF
2 = CRLF
```

- And `assert(endnl <= 2)`.
- This eliminates the `crlf` param.
- Otherwise there's an implicit coupling between `endnl` and `crlf`. if crlf=true and endnl=false, that doesn't make sense.

3N4N replied:
>     if crlf=true and endnl=false, that doesn't make sense.

But this condition may exist, can't it? In L6238, we're checking only for CRLF vs LF; the value of `endnl` doesn't enter here. What should be the value of the TriState var for this line when `endnl=false`?

justinmk replied:
there's no reason for it to exist so we should prevent it

3N4N replied:
Sadly, still confused.

There are two places where `endnl` and `crlf` are used.

```
    *len += (ptrdiff_t)strlen(tv_get_string(TV_LIST_ITEM_TV(li))) + (crlf ? 2 : 1);  // Ln6238
<snip>
  char *ret = xmalloc((size_t)(*len) + (endnl ? (crlf ? 2 : 1) : 0));                          // Ln6245
```

In the second case, I understand there are three possible values, so a tristate variable. But if `endnl` is 0, which _is_ the case with the two places this function is called from<sup>[1],[2]</sup>, how am I to know whether to use `CRLF` or `LF` in Ln6238?

I may be completely mistaking your point.


[1]: https://github.com/3N4N/neovim/blob/05d3625486a7/src/nvim/eval/funcs.c#L809
[2]: https://github.com/3N4N/neovim/blob/05d3625486a7/src/nvim/eval.c#L5691

justinmk replied:
You are right. I should have looked closer. I thought that `endnl` was controlling if the lines were joined at all.

justinmk commented on src/nvim/eval/funcs.c:
```diff
@@ -791,22 +791,29 @@ static void f_chansend(typval_T *argvars, typval_T *rettv, EvalFuncData fptr)
+#else
+  Channel *chan = find_channel(id);
+  bool crlf = (chan != NULL && chan->term) ? true: false;
+#endif
```
out of scope for this PR, but I wonder if given https://github.com/neovim/neovim/issues/16957 and https://github.com/neovim/neovim/issues/6751 we might need to check 'shell' or something to decide chansend() line endings.

3N4N replied:
I'll try manual test in mintty shell.

3N4N replied:
Tried in MSys2 and MSys2/MinGW64 shells in mintty shell. Works. But that's because the shells tested all require CRLF line-ending. So you're right. We need a mapping table or similar to decide based on `shell`.

justinmk replied:
on Windows, all shells should be able to handle CRLF. This was more of a "nice to have" idea. Should wait until we have an actual use case.

justinmk commented on test/functional/terminal/channel_spec.lua:
```diff
@@ -92,3 +93,18 @@ describe('terminal channel is closed and later released if', function()
   end)
 end)
+
+if iswin() then
```
not really any reason for this test to only run on Windows. We just would only use `sh` for non-Windows systems.

3N4N replied:
Done

justinmk commented on test/functional/terminal/channel_spec.lua:
```diff
@@ -92,3 +93,20 @@ describe('terminal channel is closed and later released if', function()
+  clear()
+  local screen = Screen.new(100, 20)
+  screen:attach()
+  local shells = {'sh'}
+  if iswin() then
+    shells = {'cmd.exe', 'pwsh.exe -nop','powershell.exe -nop'}
+  end
```
ternary :)
```suggestion
  local shells = iswin() and {'cmd.exe', 'pwsh.exe -nop','powershell.exe -nop'} or {'sh'}
```

3N4N committed:
fix(chansend): sending lines to terminal in reverse order on Windows #19315

Problem: `chansend()` on Windows sends lines in reverse order.
Cause: Using \n instead of \r\n for newlines on Windows.
Solution: on Windows, use CRLF newline characters.

Fixes #18501

justinmk committed:
fixup! fix(chansend): sending lines to terminal in reverse order on Windows #19315

justinmk commented:
rebased, waiting for ci
